### PR TITLE
Mips update

### PR DESCRIPTION
**Pull request by wargio**
# squash me
 

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Requires https://github.com/capstone-engine/capstone/pull/2410

### CODE REVIEW

**Review by XVilka — CHANGES_REQUESTED** [review]
I recommend in the same PR to extract ESIL uplifting into a separate file as well, will be easier to add RzIL afterwards

File: librz/arch/p/arch_mips.c [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+		return false;
+	}
+#endif /* CS_NEXT_VERSION */
+	return true;
+}
+#undef return_on_cpu
+
+#include "analysis/analysis_mips.c"
+#include "asm/asm_mips.c"
+#include "parse/parse_mips_pseudo.c"
```
Comment (Rot127) [inline librz/arch/p/arch_mips.c]: This is just until RzArch is done, right?

File: librz/arch/p/asm/asm_mips.c [hunk 2]
```diff
@@ -3,32 +3,28 @@
 
 #include <rz_asm.h>
 #include <rz_lib.h>
-#include <mips/mips_assembler.h>
+#include <mips/mips_internal.h>
+#include "capstone.h"
```
Comment (Rot127) [inline librz/arch/p/asm/asm_mips.c]:
```suggestion
#include <capstone/capstone.h>
```

File: librz/arch/p/asm/asm_mips.c [hunk 3]
```diff
@@ -3,32 +3,28 @@
...
 #include "cs_helper.h"
+#include "rz_util/rz_log.h"
```
Comment (Rot127) [inline librz/arch/p/asm/asm_mips.c]:
```suggestion
#include <rz_util/rz_log.h>
```

File: librz/core/cbin.c [hunk 4]
```diff
@@ -13,6 +13,7 @@
 
 #include "../bin/dwarf/dwarf_private.h"
 #include "core_private.h"
+#include "rz_util/rz_str.h"
```
Comment (Rot127) [inline librz/core/cbin.c]:
```suggestion
#include <rz_util/rz_str.h>
```

File: librz/arch/p/arch_mips.c (hunk 1 above)
Comment (wargio) [reply librz/arch/p/arch_mips.c]: yes

File: librz/arch/p/asm/asm_mips.c (hunk 2 above)
Comment (wargio) [reply librz/arch/p/asm/asm_mips.c]: that was added by the IDE, i have to change it

Comment (Rot127) [reply librz/arch/p/asm/asm_mips.c]: Yeah, happens to me all the time as well. Annoying.

File: librz/arch/isa/xtensa/xtensa.c [hunk 5]
```diff
@@ -27,16 +27,16 @@ bool xtensa_open(XtensaContext *ctx, const char *cpu, bool big_endian) {
 		return false;
 	}
 	cs_mode mode = big_endian ? CS_MODE_BIG_ENDIAN : CS_MODE_LITTLE_ENDIAN;
-	if (RZ_STR_ISEMPTY(cpu)) {
-		mode |= xtensa_cpu_modes[0].mode;
-	} else {
+	cs_mode mcpu = xtensa_cpu_modes[0].mode;
+	if (RZ_STR_ISNOTEMPTY(cpu)) {
```
Comment (XVilka) [inline librz/arch/isa/xtensa/xtensa.c]: You could just rebase as it was fixed in current `dev` IIRC

**Review by XVilka — CHANGES_REQUESTED** [review]

**Comment by XVilka** [comment]
```
 [XX] 220 ms   db/esil/mips_32 mul s0, a0, v0
RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -qc 'e asm.arch=mips
e asm.bits=32
wx 02808270
ar > /dev/null
ar a0=21
ar v0=12
aes
ar s0
ar > /dev/null
ar a0=0x44556677
ar v0=0x00112233
aeip
aes
ar s0
' malloc://0x200
-- stdout
--- expected
+++ actual
@@ -1,2 +1,2 @@
-s0 = 0x000000fc
-s0 = 0xc28637b5
+s0 = 0x00000000
+s0 = 0x00000000
```

**Review by XVilka — CHANGES_REQUESTED** [review]
```
[XX]   1 secs db/json/json2 <json> iAj
```
with ASAN enabled

**Comment by XVilka** [comment]
Please also rebase on top of `dev` to fix the DWARF Unicode test

**Comment by wargio** [comment]
> ```
> [XX]   1 secs db/json/json2 <json> iAj
> ```
> 
> with ASAN enabled

cannot repro, but added some new changes where i actually modified the code

File: librz/arch/p/asm/asm_mips_cs.c [hunk 6]
```diff
@@ -52,19 +52,27 @@ static int mips_disassemble(RzAsm *a, RzAsmOp *op, const ut8 *buf, int len) {
 	n = cs_disasm(ctx->handle, (ut8 *)buf, len, a->pc, 1, &insn);
 	if (n < 1) {
 		rz_asm_op_set_asm(op, "invalid");
-		op->size = 4;
+#if CS_NEXT_VERSION < 6
+		op->size = mode & CS_MODE_MICRO ? 2 : 4;
+#else
+		op->size = mode & (CS_MODE_MICRO | CS_MODE_NANOMIPS | CS_MODE_MIPS16) ? 2 : 4;
```
Comment (XVilka) [inline librz/arch/p/asm/asm_mips_cs.c:58]: Since you added Nanomips, please add tests for it too.

Comment (XVilka) [reply librz/arch/p/asm/asm_mips_cs.c:58]: Same for micromips

**Review by XVilka — CHANGES_REQUESTED** [review]

**Review by XVilka — APPROVED** [review]

**Comment by XVilka** [comment]
```
librz_bin.so.0.8.0.p/dwarf_dwarf.c.o -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,-soname,librz_bin.so.0.8 '-Wl,-rpath,$ORIGIN/../util:$ORIGIN/../demangler:$ORIGIN/../socket:$ORIGIN/../magic:$ORIGIN/../hash:$ORIGIN/../syscall:$ORIGIN/../cons:$ORIGIN/../io:$ORIGIN/../crypto:$ORIGIN/../type' -Wl,-rpath-link,/home/runner/work/rizin/rizin/build/subprojects/rizin/librz/util -Wl,-rpath-link,/home/runner/work/rizin/rizin/build/subprojects/rizin/librz/demangler -Wl,-rpath-link,/home/runner/work/rizin/rizin/build/subprojects/rizin/librz/socket -Wl,-rpath-link,/home/runner/work/rizin/rizin/build/subprojects/rizin/librz/magic -Wl,-rpath-link,/home/runner/work/rizin/rizin/build/subprojects/rizin/librz/hash -Wl,-rpath-link,/home/runner/work/rizin/rizin/build/subprojects/rizin/librz/syscall -Wl,-rpath-link,/home/runner/work/rizin/rizin/build/subprojects/rizin/librz/cons -Wl,-rpath-link,/home/runner/work/rizin/rizin/build/subprojects/rizin/librz/io -Wl,-rpath-link,/home/runner/work/rizin/rizin/build/subprojects/rizin/librz/crypto -Wl,-rpath-link,/home/runner/work/rizin/rizin/build/subprojects/rizin/librz/type -Wl,--start-group subprojects/lz4-1.9.4/liblz4.a subprojects/zstd-1.5.5/lib/libzstd.a subprojects/rizin/subprojects/yxml/libyxml.a subprojects/rizin/librz/util/librz_util.so.0.8.0 subprojects/rizin/librz/demangler/librz_demangler.so.0.8.0 subprojects/rizin/librz/socket/librz_socket.so.0.8.0 subprojects/rizin/librz/magic/librz_magic.so.0.8.0 subprojects/rizin/librz/hash/librz_hash.so.0.8.0 subprojects/rizin/librz/syscall/librz_syscall.so.0.8.0 subprojects/rizin/librz/cons/librz_cons.so.0.8.0 subprojects/rizin/librz/io/librz_io.so.0.8.0 subprojects/rizin/librz/type/librz_type.so.0.8.0 subprojects/libmspack/liblibmspack.a -Wl,--end-group -pthread
/usr/bin/ld: subprojects/rizin/librz/bin/librz_bin.so.0.8.0.p/format_elf_elf_info.c.o: in function `arch_is_mips':
/home/runner/work/rizin/rizin/build/../.github/subproject_test/subprojects/rizin/librz/bin/format/elf/elf_info.c:736: undefined reference to `arch_is_nanomips'
/usr/bin/ld: subprojects/rizin/librz/bin/librz_bin.so.0.8.0.p/format_elf_elf_info.c.o: in function `Elf32_rz_bin_elf_get_cpu':
/home/runner/work/rizin/rizin/build/../.github/subproject_test/subprojects/rizin/librz/bin/format/elf/elf_info.c:1625: undefined reference to `arch_is_nanomips'
/usr/bin/ld: subprojects/rizin/librz/bin/librz_bin.so.0.8.0.p/format_elf_elf64_info.c.o: in function `arch_is_mips':
/home/runner/work/rizin/rizin/build/../.github/subproject_test/subprojects/rizin/librz/bin/format/elf/elf_info.c:736: undefined reference to `arch_is_nanomips'
/usr/bin/ld: subprojects/rizin/librz/bin/librz_bin.so.0.8.0.p/format_elf_elf64_info.c.o: in function `Elf64_rz_bin_elf_get_cpu':
/home/runner/work/rizin/rizin/build/../.github/subproject_test/subprojects/rizin/librz/bin/format/elf/elf_info.c:1625: undefined reference to `arch_is_nanomips'
collect2: error: ld returned 1 exit status
```

**Pull request merged**